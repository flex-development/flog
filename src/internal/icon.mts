/**
 * @file Internal - icon
 * @module log/internal/icon
 */

/**
 * Get an icon based on unicode support.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {string} icon
 *  Unicode icon
 * @param {string} fallback
 *  Icon to use if unicode is not supported
 * @param {boolean} unicode
 *  Unicode supported?
 * @return {string}
 *  Icon string
 */
function icon(
  this: void,
  icon: string,
  fallback: string,
  unicode: boolean
): string {
  return unicode ? icon : fallback
}

export default icon
