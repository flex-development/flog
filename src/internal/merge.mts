/**
 * @file Internal - merge
 * @module log/internal/merge
 */

import { ksort, isObjectPlain as pojo } from '@flex-development/tutils'

/**
 * Merge one or more objects into a single object.
 *
 * Source objects are applied from left to right. Subsequent sources overwrite
 * property assignments of previous sources.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Record<PropertyKey, any>} target
 *  Target object
 * @param {Record<PropertyKey, any>[]} sources
 *  Source object(s)
 * @return {Record<PropertyKey, any>}
 *  Merged object
 */
function merge(
  this: void,
  target: Record<PropertyKey, any>,
  ...sources: Record<PropertyKey, any>[]
): Record<PropertyKey, any> {
  return ksort(sources.reduce<Record<PropertyKey, any>>((acc, source) => {
    return [
      ...Object.getOwnPropertySymbols(source),
      ...Object.getOwnPropertyNames(source)
    ].reduce((target, property) => {
      /**
       * Source object value.
       *
       * @var {any} value
       */
      let value: any = source[property]

      if (Object.prototype.hasOwnProperty.call(target, property)) {
        /**
         * Target object value.
         *
         * @const {any} targetValue
         */
        const targetValue: any = target[property]

        if (pojo(targetValue) && pojo(value)) {
          value = merge(targetValue, value)
        }
      }

      return target[property] = value, target
    }, acc)
  }, target))
}

export default merge
