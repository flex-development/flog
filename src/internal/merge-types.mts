/**
 * @file Internal - mergeTypes
 * @module log/internal/mergeTypes
 */

import logLevels from '#enums/log-levels'
import logTypes from '#enums/log-types'
import icon from '#internal/icon'
import merge from '#internal/merge'
import type { InputLogObject, LogType } from '@flex-development/log'

/**
 * Create a log type to configuration map.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Partial<Record<LogType, InputLogObject>> | null | undefined} types
 *  User log type to config map
 * @param {boolean} unicode
 *  Unicode supported?
 * @return {Record<LogType, InputLogObject>}
 *  Log type to configuration map
 */
function mergeTypes(
  this: void,
  types: Partial<Record<LogType, InputLogObject>> | null | undefined,
  unicode: boolean
): Record<LogType, InputLogObject> {
  /**
   * Checkmark icon.
   *
   * @const {string} check
   */
  const check: string = icon('✔', '√', unicode)

  /**
   * Failure icon.
   *
   * @const {string} x
   */
  const x: string = icon('✖', '×', unicode)

  /**
   * Default log type to configuration map.
   *
   * @const {Record<LogType, InputLogObject>} defaults
   */
  const defaults: Record<LogType, InputLogObject> = {
    [logTypes.debug]: {
      color: 'gray',
      icon: icon('⚙', 'D', unicode),
      level: logLevels.debug,
      type: logTypes.debug
    },
    [logTypes.error]: {
      color: 'red',
      icon: x,
      level: logLevels.error,
      type: logTypes.error
    },
    [logTypes.fail]: {
      color: 'red',
      icon: x,
      level: logLevels.error,
      type: logTypes.fail
    },
    [logTypes.info]: {
      color: 'cyan',
      icon: icon('ℹ', 'i', unicode),
      level: logLevels.info,
      type: logTypes.info
    },
    [logTypes.inspect]: {
      level: logLevels.log,
      type: logTypes.inspect
    },
    [logTypes.log]: {
      level: logLevels.log
    },
    [logTypes.ready]: {
      color: 'green',
      icon: check,
      level: logLevels.info,
      type: logTypes.ready
    },
    [logTypes.start]: {
      color: 'magenta',
      icon: icon('◐', 'o', unicode),
      level: logLevels.info,
      type: logTypes.start
    },
    [logTypes.success]: {
      color: 'green',
      icon: check,
      level: logLevels.info,
      type: logTypes.success
    },
    [logTypes.trace]: {
      icon: icon('→', '→', unicode),
      level: logLevels.trace,
      type: logTypes.trace
    },
    [logTypes.warn]: {
      color: 'yellow',
      icon: icon('▲', '‼', unicode),
      level: logLevels.warn,
      type: logTypes.warn
    },
    [logTypes.verbose]: {
      level: logLevels.verbose,
      type: logTypes.verbose
    }
  }

  return merge(defaults, { ...types })
}

export default mergeTypes
