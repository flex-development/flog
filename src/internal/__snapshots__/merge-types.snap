// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/mergeTypes > should return log type to configuration map 1`] = `
{
  "debug": {
    "color": "gray",
    "icon": "⚙",
    "level": 4,
    "type": "debug",
  },
  "error": {
    "color": "red",
    "icon": "✖",
    "level": 0,
    "type": "error",
  },
  "fail": {
    "color": "red",
    "icon": "✖",
    "level": 0,
    "type": "fail",
  },
  "info": {
    "color": "cyan",
    "icon": "ℹ",
    "level": 3,
    "type": "info",
  },
  "inspect": {
    "level": 2,
    "type": "inspect",
  },
  "log": {
    "level": 2,
  },
  "ready": {
    "color": "green",
    "icon": "✔",
    "level": 3,
    "type": "ready",
  },
  "start": {
    "color": "magenta",
    "icon": "◐",
    "level": 3,
    "type": "start",
  },
  "success": {
    "color": "green",
    "icon": "✔",
    "level": 3,
    "type": "success",
  },
  "trace": {
    "icon": "→",
    "level": 5,
    "type": "trace",
  },
  "verbose": {
    "level": 999,
    "type": "verbose",
  },
  "warn": {
    "color": "yellow",
    "icon": "▲",
    "level": 1,
    "type": "warn",
  },
}
`;
